File: playwright-demo.mjs
// End-to-end demo:
//   1. Launch local Chromium with the MagicCDP extension already loaded
//      (--load-extension), with the CDP port open on localhost. This step is
//      what we are saying any user can already do — no MagicCDP client code is
//      involved.
//   2. Spawn the standalone MagicCDPBridge proxy in front of that browser.
//   3. Connect Playwright via vanilla `chromium.connectOverCDP(bridgeUrl)`.
//   4. From an unmodified Playwright CDPSession, send Magic.* / Custom.*
//      commands and receive Custom.* events.

import { spawn } from "node:child_process";
import { mkdtemp, rm } from "node:fs/promises";
import net from "node:net";
import { tmpdir } from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { setTimeout as sleep } from "node:timers/promises";
import { chromium } from "playwright";

const ROOT = path.dirname(fileURLToPath(import.meta.url));
const EXT = path.join(ROOT, "extension");
const CHROME = process.env.CHROME_PATH || "/opt/pw-browsers/chromium-1194/chrome-linux/chrome";

async function freePort() {
  const s = net.createServer();
  await new Promise(r => s.listen(0, "127.0.0.1", r));
  const { port } = s.address();
  await new Promise(r => s.close(r));
  return port;
}

async function waitForUrl(url, timeoutMs = 8000) {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    try { const r = await fetch(url); if (r.ok) return await r.json(); } catch {}
    await sleep(60);
  }
  throw new Error(`timeout waiting for ${url}`);
}

async function main() {
  const events = [];
  let chromeProc, bridgeProc, browser, profileDir;

  try {
    // 1. launch Chromium with the MagicCDP extension preloaded
    profileDir = await mkdtemp(path.join(tmpdir(), "magic-cdp-pw."));
    const chromePort = await freePort();
    console.log("== launching Chromium with MagicCDP extension at port", chromePort);
    chromeProc = spawn(CHROME, [
      "--headless=new",
      "--no-sandbox",
      "--disable-gpu",
      "--enable-unsafe-extension-debugging",
      "--remote-allow-origins=*",
      `--remote-debugging-port=${chromePort}`,
      `--user-data-dir=${profileDir}`,
      `--load-extension=${EXT}`,
      "about:blank",
    ], { stdio: ["ignore", "ignore", "ignore"] });
    const chromeVersion = await waitForUrl(`http://127.0.0.1:${chromePort}/json/version`);
    console.log("chromium up:", chromeVersion.Browser);

    // 2. spawn MagicCDPBridge as a separate process in front of Chrome
    const bridgePort = await freePort();
    const bridgeUrl = `http://127.0.0.1:${bridgePort}`;
    console.log("\n== spawning MagicCDPBridge at", bridgeUrl);
    bridgeProc = spawn(process.execPath, [
      path.join(ROOT, "bridge.mjs"),
      "--upstream", `http://127.0.0.1:${chromePort}`,
      "--port", String(bridgePort),
    ], { stdio: ["ignore", "inherit", "inherit"] });
    await waitForUrl(`${bridgeUrl}/json/version`);

    // 3. connect Playwright via vanilla connectOverCDP
    console.log("\n== connecting Playwright via connectOverCDP through bridge ==");
    browser = await chromium.connectOverCDP(bridgeUrl);
    console.log("playwright connected. contexts:", browser.contexts().length);

    const ctx = browser.contexts()[0] ?? await browser.newContext();
    const page = ctx.pages()[0] ?? await ctx.newPage();
    console.log("page url:", page.url());

    // Vanilla Playwright CDPSession scoped to a page target.
    const session = await ctx.newCDPSession(page);

    console.log("\n== prove standard CDP still works through the bridge ==");
    console.log("Browser.getVersion ->", await session.send("Browser.getVersion"));

    console.log("\n== Magic.evaluate from a vanilla Playwright CDPSession ==");
    const info = await session.send("Magic.evaluate", {
      expression: "async () => ({ extensionId: chrome.runtime.id, swUrl: chrome.runtime.getURL('service_worker.js') })",
    });
    console.log("Magic.evaluate result ->", info);

    console.log("\n== Magic.addCustomEvent + session.on('Custom.demo', ...) ==");
    await session.send("Magic.addCustomEvent", { name: "Custom.demo" });
    session.on("Custom.demo", payload => {
      console.log("PLAYWRIGHT RECEIVED Custom.demo ->", payload);
      events.push(payload);
    });

    console.log("\n== Magic.addCustomCommand: Custom.echo emits Custom.demo ==");
    await session.send("Magic.addCustomCommand", {
      name: "Custom.echo",
      expression: "async (params, { cdp }) => { await cdp.emit('Custom.demo', { echo: params.value, ts: Date.now() }); return { ok: true, echoed: params.value }; }",
    });

    console.log("\n== Custom.echo (custom command) sent from Playwright ==");
    const echo1 = await session.send("Custom.echo", { value: "hello-from-playwright" });
    console.log("Custom.echo result ->", echo1);
    const echo2 = await session.send("Custom.echo", { value: "second-roundtrip" });
    console.log("Custom.echo result ->", echo2);

    const waitDeadline = Date.now() + 3000;
    while (events.length < 2 && Date.now() < waitDeadline) await sleep(20);

    console.log("\n== summary ==");
    console.log("events received:", events);
    if (events.length < 2) throw new Error(`expected >=2 Custom.demo events, got ${events.length}`);
    if (echo1.echoed !== "hello-from-playwright") throw new Error("echo1 did not roundtrip");
    if (echo2.echoed !== "second-roundtrip") throw new Error("echo2 did not roundtrip");

    console.log("\nSUCCESS: Playwright CDPSession sent Magic commands and received Magic events through the bridge.");
  } finally {
    try { await browser?.close(); } catch {}
    try { bridgeProc?.kill("SIGTERM"); } catch {}
    try { chromeProc?.kill("SIGTERM"); } catch {}
    if (profileDir) await rm(profileDir, { recursive: true, force: true }).catch(() => {});
  }
}

main().catch(e => { console.error("DEMO FAILED:", e); process.exitCode = 1; });

File: playwright-demo.py
"""End-to-end demo: stock Playwright Python connecting through the same JS
MagicCDPBridge that the Node demo uses, against a real local Chromium.

What this proves:
  - chromium.connect_over_cdp(...) works against the JS proxy unchanged.
  - cdp_session.send('Magic.*' / 'Custom.*', ...) works from Python.
  - cdp_session.on('Custom.*', handler) receives binding events from the SW.

The chrome + proxy lifecycle here intentionally mirrors playwright-demo.mjs:
  - Launch upstream Chromium with --load-extension (workaround for the
    Playwright-bundled chromium build that doesn't implement
    Extensions.loadUnpacked over CDP).
  - Spawn `node proxy.mjs --upstream <url> --port <free>`.
  - Connect Playwright Python via connect_over_cdp through the proxy.

Nothing about playwright python needed any patching.
"""

import json
import os
import socket
import subprocess
import sys
import tempfile
import time
import urllib.request
from pathlib import Path

from playwright.sync_api import sync_playwright

ROOT = Path(__file__).resolve().parent
EXTENSION_PATH = ROOT / "extension"
CHROME = os.environ.get("CHROME_PATH", "/opt/pw-browsers/chromium-1194/chrome-linux/chrome")


def free_port() -> int:
    s = socket.socket()
    s.bind(("127.0.0.1", 0))
    port = s.getsockname()[1]
    s.close()
    return port


def wait_for_url(url: str, timeout_s: float = 8.0) -> dict:
    deadline = time.monotonic() + timeout_s
    while time.monotonic() < deadline:
        try:
            with urllib.request.urlopen(url, timeout=0.5) as response:
                return json.loads(response.read())
        except Exception:
            time.sleep(0.05)
    raise RuntimeError(f"timeout waiting for {url}")


def main() -> int:
    events: list[dict] = []
    chrome_proc = None
    proxy_proc = None
    profile_dir = tempfile.mkdtemp(prefix="magic-cdp-py.")

    try:
        # 1. launch upstream Chromium with the extension preloaded
        chrome_port = free_port()
        print(f"== launching upstream Chromium at port {chrome_port}")
        chrome_proc = subprocess.Popen(
            [
                CHROME,
                "--headless=new",
                "--no-sandbox",
                "--disable-gpu",
                "--enable-unsafe-extension-debugging",
                "--remote-allow-origins=*",
                "--no-first-run",
                "--no-default-browser-check",
                f"--remote-debugging-port={chrome_port}",
                f"--user-data-dir={profile_dir}",
                f"--load-extension={EXTENSION_PATH}",
                "about:blank",
            ],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
        chrome_url = f"http://127.0.0.1:{chrome_port}"
        version = wait_for_url(f"{chrome_url}/json/version")
        print(f"upstream chromium up: {version['Browser']}")

        # 2. spawn the JS proxy in front of that Chromium
        proxy_port = free_port()
        proxy_url = f"http://127.0.0.1:{proxy_port}"
        print(f"\n== spawning MagicCDPBridge at {proxy_url}")
        proxy_proc = subprocess.Popen(
            ["node", str(ROOT / "proxy.mjs"), "--upstream", chrome_url, "--port", str(proxy_port)],
            stdout=sys.stdout,
            stderr=sys.stderr,
        )
        wait_for_url(f"{proxy_url}/json/version")

        # 3. connect Playwright Python via connect_over_cdp
        print("\n== connecting Playwright Python via connect_over_cdp through proxy ==")
        with sync_playwright() as p:
            browser = p.chromium.connect_over_cdp(proxy_url)
            print(f"playwright connected. contexts: {len(browser.contexts)}")
            ctx = browser.contexts[0] if browser.contexts else browser.new_context()
            page = ctx.pages[0] if ctx.pages else ctx.new_page()
            print(f"page url: {page.url}")

            # vanilla Playwright Python CDPSession scoped to a page target
            session = ctx.new_cdp_session(page)

            print("\n== prove standard CDP still works through the proxy ==")
            ver = session.send("Browser.getVersion")
            print(f"Browser.getVersion -> {ver}")

            print("\n== Magic.evaluate from a vanilla Playwright Python CDPSession ==")
            info = session.send("Magic.evaluate", {
                "expression": "async () => ({ extensionId: chrome.runtime.id, swUrl: chrome.runtime.getURL('service_worker.js') })",
            })
            print(f"Magic.evaluate result -> {info}")

            print("\n== Magic.addCustomEvent + session.on('Custom.demo', ...) ==")
            session.send("Magic.addCustomEvent", {"name": "Custom.demo"})

            def on_demo(payload):
                print(f"PYTHON RECEIVED Custom.demo -> {payload}")
                events.append(payload)
            session.on("Custom.demo", on_demo)

            print("\n== Magic.addCustomCommand: Custom.echo emits Custom.demo ==")
            session.send("Magic.addCustomCommand", {
                "name": "Custom.echo",
                "expression": (
                    "async (params, { cdp }) => { "
                    "  await cdp.emit('Custom.demo', { echo: params.value, ts: Date.now() }); "
                    "  return { ok: true, echoed: params.value }; "
                    "}"
                ),
            })

            print("\n== Custom.echo (custom command) sent from Playwright Python ==")
            echo1 = session.send("Custom.echo", {"value": "hello-from-python"})
            print(f"Custom.echo result -> {echo1}")
            echo2 = session.send("Custom.echo", {"value": "second-roundtrip-py"})
            print(f"Custom.echo result -> {echo2}")

            deadline = time.monotonic() + 3
            while len(events) < 2 and time.monotonic() < deadline:
                time.sleep(0.02)

            print("\n== summary ==")
            print(f"events received: {events}")
            assert len(events) >= 2, f"expected >=2 Custom.demo events, got {len(events)}"
            assert echo1["echoed"] == "hello-from-python", "echo1 did not roundtrip"
            assert echo2["echoed"] == "second-roundtrip-py", "echo2 did not roundtrip"

            print("\nSUCCESS: Playwright Python CDPSession sent Magic commands and received Magic events through the proxy.")
            browser.close()
        return 0
    finally:
        if proxy_proc is not None:
            proxy_proc.terminate()
            try: proxy_proc.wait(timeout=3)
            except Exception: proxy_proc.kill()
        if chrome_proc is not None:
            chrome_proc.terminate()
            try: chrome_proc.wait(timeout=3)
            except Exception: chrome_proc.kill()
        try: subprocess.run(["rm", "-rf", profile_dir], check=False)
        except Exception: pass


if __name__ == "__main__":
    sys.exit(main())
